Gazebo Ros2 Ardupilot Camera Support

mhl787156 opened the pull request:
This PR introduces camera support for iris ardupilot in particular, but also allows for generic cameras to be added to model URDFs. 
This changes include the following:

- `starling-sim-base-core` gazebo install updated to download `ros-foxy-gazebo-pkgs` as it contains the plugins required for sensor simulation
- A script `setup_display.sh` which configures a virtual framebuffer so that the cameras can virtually render 
- Iris-ap now includes the gimble arm and meshes, where the camera has now been configured with a camera controller in the ros2 style: [Ros2 migration guide](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-Camera)

Starts to address #88 

Future Improvements:
- `setup_display.sh` should be able to setup from a user defined choice of nvidia or virtual framebuffer

mhl787156 commented:
Added xacro parsing of the SDF file. The sdf parameters are now editable through the launch file and environment variables:
```
    <!-- Xacro launch the gimball small 2d for the iris_with_ardupilot model -->
    <executable unless="$(var sim_only)" 
        cmd="xacro model.sdf.xacro
              ros_namespace:=$(env VEHICLE_NAMESPACE)
              camera_name:=$(env CAMERA_NAME camera)
              camera_height:=$(env CAMERA_HEIGHT 480)
              camera_width:=$(env CAMERA_WIDTH 640)
              gimbal_initial_angle:=$(env GIMBAL_INITIAL_ANGLE 0.785)
              -o model.sdf
              "
        cwd="/ardupilot_gazebo/models/gimbal_small_2d/"
        output="screen"
    /> 
```

VEHICLE_NAMESPACE is now set by default in the arduino-gazebo container to `vehicle_${AP_SYSID}` so `AP_SYSID` must be set when launching a vehicle (otherwise an error will be raised)

rob-clarke commented:
The xacro stuff here will likely conflict with the changes I've made to support the non-network host mode (#117).  I suggest we merge that first as it's simpler then restructure here where needed.

mhl787156 commented:
Alright, then feel free to merge #117 and I'll come back and fix this one if there's any issues.

mhl787156 commented:
@rob-clarke 

1. Should AP_SITL_ADDDRESS/ AP_SITL_HOST check be moved into the ardupilot-gazebo container? For this branch I needed access to vehicle namespace and AP_SYSID so we have a `sim_setup.sh` script in base/ardupilot specifically for the ardupilot-gazebo container
2. I will move the gimbal xacro to the iris_setup.sh from the launch file, it's a better solution.

rob-clarke commented:
Yes, we could move the underlying environment variable logic. It was mainly there because it was model specific xacro that needed modifying so I felt it fit better at the vehicle layer. We would need to figure out if we need control over ordering of the `ros.env.d` scripts if we moved it where as now it definitely happens before the xacro call.
Could go the udev rules route and prepend `00_` etc for ordering.

mhl787156 commented:
@rob-clarke I have moved some things around, but currently having issues loading iris_with_ardupilot xacro model into gazebo. It only appears to load the iris_with_standoffs model from /ardutpilot_gazebo/models. 

The xacro should overwrite the model.sdf in /ardupilot_gazebo/models/iris_with_ardupilot, so i'm thinking something is not quite write about the generated sdf file. Everything looks about right though...

mhl787156 commented:
Update, I have isolated it to the `<listen_addr>$(arg listen_addr)</listen_addr>` line in `model.sdf.xacro`. With this line, gzserverfor model path `/ardupilot_gazebo/models` only has `iris_with_standoffs`. Without, it confirms that `/ardupilot_gazebo/models` is also overriding `iris_with_ardupilot`. 

I have checked the source so `listen_addr` is read in the libArduPilotPlugin.so. After turning on verbose mode, it also looks whatever happening is making it fail silently. (None of the logs i can find so anything). 

When I was debugging the gimbal model, I also faced some silent failures when the xacro file generated a 'invalid' sdf file. I have a feeling something similar might be happening, but I cannot work out why that particular line is making it invalid!

mhl787156 commented:
Found a solution! For some reason `xacro:args` does not work. Replaced with `xacro:property` in a similar manner to how the gimbal model.sdf.xacro was implemented. 

Gimbal now uses std_msgs/msg/Float32.msg.I am aware it is being deprecated, but could not find a better topic short of doing proper Pose. Pose2D is not available on foxglove.

Also updated Docs.

I believe this is ready to merge on approved review.

rob-clarke commented on simulator/vehicles/iris-ap/Dockerfile:
```diff
@@ -6,13 +6,19 @@ FROM ${REGISTRY}uobflightlabstarling/starling-sim-ardupilot-gazebo:${VERSION}
+    && export CMAKE_PREFIX_PATH=$AMENT_PREFIX_PATH:$CMAKE_PREFIX_PATH \
+    && colcon build --cmake-force-configure \
+    && rm -r build \
+    && echo 'export GAZEBO_PLUGIN_PATH=/ros_ws/install/gimbal_plugin/lib:${GAZEBO_PLUGIN_PATH}' >> /ros.env.d/02_gimbal_small_2d/setup.bash
```
Could this not just be part of setup.bash to start with?

mhl787156 replied:
Yes it could, I just wanted to add it here otherwise it might be a bit 'magical'? I could add it if thats better?

rob-clarke commented on simulator/base/ardupilot/sim_setup.sh:
```diff
@@ -0,0 +1,46 @@
+else
+    export VEHICLE_NAMESPACE="vehicle_$AP_SYSID"
+    echo "VEHICLE_NAMESPACE not set, default to auto generated default based on ardupilot sysid target ($AP_SYSID) of $VEHICLE_NAMESPACE"
+fi
```
As I understand it, getting the vehicle namespace is only needed to setup the topics for the camera output?
While not an issue now, how does this scale if we wanted to spawn multiple AP vehicles?

mhl787156 replied:
Yeah its only used for camera topic output. For multiple vehicles, hopefully in the same way that the PX4 side deals with it. IIRC many of the kubernetes deployments assign VEHICLE_NAMESPACE directly.

rob-clarke commented on simulator/base/core/setup_display.sh:
```diff
@@ -0,0 +1,15 @@
+set -e 
+echo "--------------- Start setup_display.sh ----------------------"
+
+if [ "$ENABLE_VIRTUAL_FRAMEBUFFER" = true ]; then
```
Is this envvar documented?

mhl787156 replied:
Adding now!

rob-clarke commented on simulator/vehicles/iris-ap/iris.launch.xml:
```diff
@@ -5,14 +5,15 @@
     <include unless="$(var spawn_only)" file="/ros_ws/launch/gzweb.launch.xml" />
-    
+
+    <!-- Launch spawn entity after deplay to ensure sdf has been generated -->
```
Is this still valid? SDF is now generated in the as part of the entrypoint scripts.

mhl787156 replied:
The comment just needed updating, still need the launch to spawn entity!

rob-clarke reviewed (changes requested):
Only main thing is to add some documentation for the `ENABLE_VIRTUAL_FRAMEBUFFER` environment variable. I had a brief look but couldn't find any.

mhl787156 commented:
Note! Currently looking at the rosbridge streaming, they all assume some level of compressed image transport. Do we need to add compression topic?

rob-clarke commented:
I'd worry about the streaming in a separate PR. Just get this in for now
